contracts/AccountBalance.sol

```diff
@@ -512,6 +532,36 @@ contract AccountBalance is IAccountBalance, BlockContext, ClearingHouseCallee, A
         return (netQuoteBalance, pendingFee);
     }
 
+    function _getMarkPrice(address baseToken) internal view virtual returns (uint256) {
+        // Use index twap:
+        //   1. For backward compatible, returns index twap when not switch to mark price yet.
+        //   2. For paused market, returns index twap as mark price.
+        if (_marketRegistry == address(0) || !IBaseToken(baseToken).isOpen()) {
+            return _getIndexPrice(baseToken, IClearingHouseConfig(_clearingHouseConfig).getTwapInterval());
+        }
+
+        uint256 marketPrice = _getMarketPrice(baseToken, 0);
+        uint256 marketTwap = _getMarketPrice(baseToken, _marketTwapInterval);
+        int256 premium =
+            _getMarketPrice(baseToken, _premiumInterval).toInt256().sub(
+                _getIndexPrice(baseToken, _premiumInterval).toInt256()
+            );
+        uint256 indexPremium = _getIndexPrice(baseToken, 0).toInt256().add(premium).toUint256();
+        return PerpMath.findMedianOfThree(marketPrice, marketTwap, indexPremium);
+    }
+
+    function _getMarketPrice(address baseToken, uint32 twapInterval) internal view returns (uint256) {
+        return
+            UniswapV3Broker
+                .getSqrtMarkTwapX96(IMarketRegistry(_marketRegistry).getPool(baseToken), twapInterval)
+                .formatSqrtPriceX96ToPriceX96()
+                .formatX96ToX10_18();
+    }
+
+    function _getIndexPrice(address baseToken, uint32 twapInterval) internal view returns (uint256) {
+        return IIndexPrice(baseToken).getIndexPrice(twapInterval);
```

Comment:
Do you know the reason why we have `IIndexPrice` and `IBaseToken`?

Reply:
I'm not sure what's the initial purpose to separate two interfaces. I think we can merge IIndexPrice into IBaseToken.
@Wraecca do you know this?

Reply:
sorry didn't setup my notification properly after migrating to this repo

in the beginning IBaseToken did not expose to AccountBalance
after v1.1.0 ( close market ) AccountBalance start to access IBaseToken, so it's fine to use IBaseToken instead of IIndexPrice in AccountBalance now